zerver/lib/retention.py

```diff
@@ -51,97 +66,106 @@ def move_expired_messages_to_archive(realm: Realm) -> None:
             AND zerver_archivedattachment.id IS NULL
        GROUP BY zerver_attachment.id
     """
     assert realm.message_retention_days is not None
-    check_date = timezone_now() - timedelta(days=realm.message_retention_days)
-    move_expired_rows(Attachment, query, realm_id=realm.id, check_date=check_date.isoformat())
+    move_expired_rows(Attachment, query, realm_id=realm.id,
+                      message_ids=ids_list_to_sql_query_format(msg_ids))
 
 
-def move_expired_attachments_message_rows_to_archive(realm: Realm) -> None:
+def move_expired_attachments_message_rows_to_archive(realm: Realm, msg_ids: List[int]) -> None:
+    if not msg_ids:
+        return
+
     query = """
        INSERT INTO zerver_archivedattachment_messages (id, archivedattachment_id, archivedmessage_id)
        SELECT zerver_attachment_messages.id, zerver_attachment_messages.attachment_id,
            zerver_attachment_messages.message_id
        FROM zerver_attachment_messages
        INNER JOIN zerver_attachment
            ON zerver_attachment_messages.attachment_id = zerver_attachment.id
-       INNER JOIN zerver_message ON zerver_attachment_messages.message_id = zerver_message.id
        LEFT JOIN zerver_archivedattachment_messages
            ON zerver_archivedattachment_messages.id = zerver_attachment_messages.id
-       WHERE zerver_attachment.realm_id = {realm_id}
-            AND  zerver_message.pub_date < '{check_date}'
+       WHERE  zerver_attachment_messages.message_id IN {message_ids}
+            AND zerver_attachment.realm_id = {realm_id}
             AND  zerver_archivedattachment_messages.id IS NULL
     """
     assert realm.message_retention_days is not None
-    check_date = timezone_now() - timedelta(days=realm.message_retention_days)
     with connection.cursor() as cursor:
-        cursor.execute(query.format(realm_id=realm.id, check_date=check_date.isoformat()))
+        cursor.execute(query.format(realm_id=realm.id,
+                                    message_ids=ids_list_to_sql_query_format(msg_ids)))
 
 
-def delete_expired_messages(realm: Realm) -> None:
+def delete_expired_messages(realm: Realm, msg_ids: List[int]) -> None:
     removing_messages = Message.objects.filter(
-        usermessage__isnull=True, id__in=ArchivedMessage.objects.all(),
+        usermessage__isnull=True, id__in=msg_ids,
         sender__realm_id=realm.id
     )
     removing_messages.delete()
 
 
-def delete_expired_user_messages(realm: Realm) -> None:
+def delete_expired_user_messages(realm: Realm, usermsg_ids: List[int]) -> None:
+    assert realm.message_retention_days is not None
     removing_user_messages = UserMessage.objects.filter(
-        id__in=ArchivedUserMessage.objects.all(),
-        user_profile__realm_id=realm.id
+        id__in=usermsg_ids,
+        user_profile__realm_id=realm.id,
+        message__pub_date__lt=(timezone_now() - timedelta(days=realm.message_retention_days)).isoformat()
```

Comment:
What's the story with adding this line?  It seems like that should have already been checked for these IDs elsewhere.

Reply:
Oh, I missed this - yes, this pub_date check should be redundant. Removing it.